File: PetMedicalCare/BackEnd/SoftPetBusiness/src/main/java/pe/edu/pucp/softpet/bo/RecetaMedicaBo.java
package pe.edu.pucp.softpet.bo;

import java.util.ArrayList;
import java.sql.Date;
import pe.edu.pucp.softpet.dao.RecetaMedicaDao;
import pe.edu.pucp.softpet.daoImp.RecetaMedicaDaoImpl;
import pe.edu.pucp.softpet.dto.citas.CitaAtencionDto;
import pe.edu.pucp.softpet.dto.recetas.RecetaMedicaDto;

public class RecetaMedicaBo {

    private final RecetaMedicaDao recetaDao;

    public RecetaMedicaBo() {
        this.recetaDao = new RecetaMedicaDaoImpl();
    }

    public Integer insertar(int citaId, String fechaEmision, String vigenciaHasta,
            String diagnostico, String observaciones, boolean activo) {

        RecetaMedicaDto receta = new RecetaMedicaDto();

        CitaAtencionDto cita = new CitaAtencionDto();
        cita.setCitaId(citaId);

        receta.setCita(cita);
        receta.setFechaEmision(Date.valueOf(fechaEmision));
        receta.setVigenciaHasta(Date.valueOf(vigenciaHasta));
        receta.setDiagnostico(diagnostico);
        receta.setObservaciones(observaciones);
        receta.setActivo(activo);

        return this.recetaDao.insertar(receta);
    }

    public Integer modificar(int recetaId, int citaId, String fechaEmision,
            String vigenciaHasta, String diagnostico, String observaciones,
            boolean activo) {

        RecetaMedicaDto receta = new RecetaMedicaDto();

        CitaAtencionDto cita = new CitaAtencionDto();
        cita.setCitaId(citaId);

        receta.setRecetaMedicaId(recetaId);
        receta.setCita(cita);
        receta.setFechaEmision(Date.valueOf(fechaEmision));
        receta.setVigenciaHasta(Date.valueOf(vigenciaHasta));
        receta.setDiagnostico(diagnostico);
        receta.setObservaciones(observaciones);
        receta.setActivo(activo);

        return this.recetaDao.modificar(receta);
    }

    public Integer eliminar(int recetaMedicaId) {
        RecetaMedicaDto dto = new RecetaMedicaDto();
        dto.setRecetaMedicaId(recetaMedicaId);
        return this.recetaDao.eliminar(dto);
    }

    public RecetaMedicaDto obtenerPorId(int recetaMedicaId) {
        return this.recetaDao.obtenerPorId(recetaMedicaId);
    }

    public ArrayList<RecetaMedicaDto> listarTodos() {
        return this.recetaDao.listarTodos();
    }

    public RecetaMedicaDto obtenerPorIdCita(int citaId) {
        return this.recetaDao.obtenerPorIdCita(citaId);
    }
    
    // ... otros métodos ...

    public ArrayList<RecetaMedicaDto> listarPorMascotaYFecha(int mascotaId, String fecha) {
        return this.recetaDao.listarPorMascotaYFecha(mascotaId, fecha);
    }
    
    public ArrayList<RecetaMedicaDto> listarBusquedaAvanzada(String mascota, String duenio, String fechaStr, String activo) {
        java.sql.Date fechaSql = null;
        if (fechaStr != null && !fechaStr.isEmpty()) {
            try {
                fechaSql = java.sql.Date.valueOf(fechaStr);
            } catch (Exception e) {}
        }
        return this.recetaDao.listarBusquedaAvanzada(mascota, duenio, fechaSql, activo);
    }
}

File: Mis últimos pasos/SoftPetWA/SoftPetBussiness/SOAPBo/Citas/RecetaMedicaBO.cs
using SoftPetBussiness.MascotaClient;
using SoftPetBussiness.RecetaMedicaClient;
using System;
using System.Collections.Generic;
using System.Linq;

namespace SoftPetBusiness
{
    public class RecetaMedicaBO
    {
        private RecetasMedicaClient clienteSOAP;

        public RecetaMedicaBO()
        {
            this.clienteSOAP = new RecetasMedicaClient();
        }

        public int Insertar(int citaId, string fechaEmision, string vigenciaHasta, string diagnostico, string observaciones, bool activo)
        {
            return this.clienteSOAP.insertar_receta(citaId, fechaEmision, vigenciaHasta, diagnostico, observaciones, activo);
        }

        public int Modificar(int recetaId, int citaId, string fechaEmision, string vigenciaHasta, string diagnostico, string observaciones, bool activo)
        {
            return this.clienteSOAP.modificar_receta(recetaId, citaId, fechaEmision, vigenciaHasta, diagnostico, observaciones, activo);
        }

        // --- AGREGADO: Necesario para el Modal "Ver Detalle" ---
        public recetaMedicaDto ObtenerPorId(int recetaId)
        {
            return this.clienteSOAP.obtener_por_id(recetaId);
        }

        public int Eliminar(int recetaId)
        {
            return this.clienteSOAP.eliminar_receta(recetaId);
        }
        // -------------------------------------------------------

        public recetaMedicaDto ObtenerPorIdCita(int citaId)
        {
            return this.clienteSOAP.obtener_receta_por_cita(citaId);
        }

        public List<recetaMedicaDto> ListarTodos()
        {
            return this.clienteSOAP.listar_todos().ToList<recetaMedicaDto>();
        }

        public List<recetaMedicaDto> ListarPorMascotaYFecha(int mascotaId, DateTime fecha)
        {
            string fechaStr = fecha.ToString("yyyy-MM-dd");
            var lista = clienteSOAP.listar_recetas_por_mascota_y_fecha(mascotaId, fechaStr);
            return lista != null ? lista.ToList() : new List<recetaMedicaDto>();
        }

        public List<recetaMedicaDto> ListarBusquedaAvanzada(string mascota, string duenio, string fecha, string activo)
        {
            mascota = mascota ?? "";
            duenio = duenio ?? "";
            fecha = fecha ?? "";
            activo = activo ?? "";

            var lista = clienteSOAP.listar_busqueda_avanzada(mascota, duenio, fecha, activo);
            return lista != null ? lista.ToList() : new List<recetaMedicaDto>();
        }
    }
}